packages/compose/README.md

```diff
@@ -500,23 +523,46 @@ _Returns_
 
 ### useResizeObserver
 
-Hook which allows to listen to the resize event of any target element when it changes size. \_Note: `useResizeObserver` will report `null` sizes until after first render.
+Tracks a given element's size and calls `onUpdate` for all of its discrete values using a `ResizeObserver`. Pass the returned ref to the element or pass the element to the `targetElement` option directly.
 
 _Usage_
 
-```js
-const App = () => {
-	const [ resizeListener, sizes ] = useResizeObserver();
+```tsx
+const targetElementRef = useResizeObserver(
+	( resizeObserverEntries, element ) => {
+		console.log( 'Resize observer entries:', resizeObserverEntries );
+		console.log( 'Element that was measured:', element );
+	},
+	{ box: 'border-box' }
+);
+<div ref={ targetElementRef } />;
+
+// Alternatively, pass the element directly as an argument:
+const [ targetElement, setTargetElement ] = useState< HTMLElement | null >();
+useResizeObserver(
+	// ...
+	{
+		targetElement,
+		// ...
+	}
+);
+<div ref={ setTargetElement } />;
```

Comment:
Personally I think we should remove this pattern. It can probably be done using the normal pattern no? Are there any use cases where this pattern is necessary?

Reply:
We can have elements that are stored in AriaKit's `CollectionStore`. Then you retrieve the rendered element with `store.item( id ).element`. There are no refs involved, except somewhere deep in AriaKit internals.

Sometimes DOM elements are passed as props from a parent component. Here the real ref is also somewhere else, far away from the usage site.

It's true that the `targetElement` pattern can be implemented in userspace:
```js
const element = store.item( id ).element;
const elementRef = useResizeObserver( onResize );
useLayoutEffect( () => {
  elementRef( element );
  return () => elementRef( null );
}, [ element, elementRef ];
```

Reply:
This is one example, and I have another where I need to do a querySelect in an effect. This API mimics floating ui's, and makes sense because not always you have a React element you can pass a ref to.

I get that we live in React land but do not think we should forget about the underlying DOM entirely. Not everything needs to (or can) go through React.

Reply:
I think it's mostly a mindset change that I'm requesting, and maybe not even specific to this API. I'm entirely certain that there are use-cases where the pattern here is more convenient but I think in our project (all of Gutenberg) we should lean on the side of "offering the least possible APIs unless really necessary" and I think that's a mindset we should be applying everywhere. For each new API, we should ask ourselves: Can we do without it? What would be the cost?

The reason is simple: The cost of maintaining an API in Gutenberg as WordPress is just way too high. It's very different than working on regular npm packages or SPA apps...

I won't be against this pattern later if we find ourselves needing it in 3 places across our codebase but until then I would personally always lean towards "least possible APIs".

Reply:
It's already been necessary 2 times. Do we really need a third? The returned ref itself is already an instance of inversion of control, and to use the utility as suggested in those 2 instances I would need to invert back myself which already feels like a hack and hard to understand/maintain. There's nothing simpler in an abstraction than letting the user pass the element instead of returning a setter that the user must call (and then passing `null` when the element is removed).

If we really abide by trying not to add any extra API, then what I'd think unnecessary is returning a callback ref and I'd stick with the original API that just takes an element in.

So I wonder where the line is, because I feel like I'm hearing both that we should be doing X because it's more practical but that we shouldn't do Y because it's unnecessary at the same time. What's the criteria for keeping something or excluding something?